File: library/src/main/java/com/apppoweron/circularrevealbutton/container/ButtonCRevealAnimationData.java
package com.apppoweron.circularrevealbutton.container;

import android.graphics.drawable.GradientDrawable;
import android.os.Build;
import android.support.annotation.RequiresApi;
import android.widget.Button;

/**
 * This class responsible for get required data to create the circular reveal animation
 */
public class ButtonCRevealAnimationData {

    private static final String TAG = "ButtonCRevealAnimationData";

    private int buttonId;
    private int x;
    private int y;
    private Integer animColor;
    private Integer animDuration;
    private int buttonRadius;


    private ButtonCRevealAnimationData(ButtonDataBuilder builder) {
        this.buttonId = builder.buttonId;
        this.x = builder.x;
        this.y = builder.y;
        this.animColor = builder.animColor;
        this.buttonRadius = builder.buttonRadius;
        this.animDuration = builder.animDuration;
    }

    public int getButtonId() {
        return this.buttonId;
    }

    public int getX() {
        return this.x;
    }

    public int getY() {
        return this.y;
    }

    public Integer getAnimColor() {
        return this.animColor;
    }

    public int getAnimStartRadius() {
        return this.buttonRadius;
    }

    public Integer getAnimDuration() {
        return this.animDuration;
    }

    public static class ButtonDataBuilder {
        private int buttonId;
        private int x;
        private int y;
        private Integer animColor;
        private Integer animDuration;
        private int buttonRadius;


        public ButtonDataBuilder(Button button, CircularRevealContainer container) {
            this.buttonRadius = button.getHeight() / 2;
            this.buttonId = button.getId();

            this.x = button.getLeft() + button.getWidth() / 2;
            this.y = button.getTop() + this.buttonRadius;

            int[] buttonLocation = {0, 0};
            int[] containerLocation = {0, 0};
            button.getLocationInWindow(buttonLocation);
            container.getLocationInWindow(containerLocation);

            this.y = buttonLocation[1] + this.buttonRadius - containerLocation[1];
        }


        @RequiresApi(api = Build.VERSION_CODES.N)
        public ButtonDataBuilder animColor(Button button) {
            GradientDrawable drawable = (GradientDrawable) button.getBackground();
            animColor(drawable.getColor().getDefaultColor());
            return this;
        }

        public ButtonDataBuilder animColor(int color) {
            this.animColor = color;
            return this;
        }

        public ButtonDataBuilder animDuration(Integer animDuration) {
            this.animDuration = animDuration;
            return this;
        }

        public ButtonCRevealAnimationData build() {
            return new ButtonCRevealAnimationData(this);
        }
    }


}

File: library/src/main/kotlin/com/apppoweron/circularrevealbutton/container/ButtonCRevealAnimationData.kt
package com.apppoweron.circularrevealbutton.container

import android.widget.Button

/**
 * This class responsible for get required data to create the circular reveal animation
 */
class ButtonCRevealAnimationData private constructor(val buttonId: Int,
                                                     val x: Int,
                                                     val y: Int,
                                                     val animColor: Int?,
                                                     val animDuration: Int?,
                                                     val animStartRadius: Int) {


    private constructor(builder: Builder) : this(builder.buttonId, builder.x, builder.y, builder.animColor, builder.animDuration, builder.buttonRadius)

    companion object {
        private val TAG = "ButtonCRevealAnimationData"
        fun create(button: Button, container: CircularRevealContainer, init: Builder.() -> Unit) = Builder(button, container, init).build()
    }

    class Builder private constructor(button: Button, container: CircularRevealContainer) {
        val buttonId: Int
        val x: Int
        val y: Int
        val buttonRadius: Int

        var animColor: Int? = null
        var animDuration: Int? = null


        init {
            this.buttonRadius = button.height / 2
            this.buttonId = button.id

            this.x = button.left + button.width / 2

            val buttonLocation = intArrayOf(0, 0)
            val containerLocation = intArrayOf(0, 0)
            button.getLocationInWindow(buttonLocation)
            container.getLocationInWindow(containerLocation)

            this.y = buttonLocation[1] + buttonRadius - containerLocation[1]
        }

        constructor(button: Button, container: CircularRevealContainer, init: Builder.() -> Unit) : this(button, container) {
            init()
        }

        fun animColor(init: Builder.() -> Int?) = apply { animColor = init() }

        fun animDuration(init: Builder.() -> Int?) = apply { animDuration = init() }

        fun build() = ButtonCRevealAnimationData(this)
    }

}
